keras/src/models/sequential.py

```diff
@@ -115,7 +115,8 @@ def add(self, layer, rebuild=True):
                 f"add a different Input layer to it."
             )
 
-        self._layers.append(layer)
+        # append will not trigger __setattr__ for tracking purpose.
+        self._layers = self._layers + [layer]
```

Comment:
Does this indicate that the tracking will fail if we add a stateful object without `__setattr__`?
Or is this a special case for `Sequential` model?

Reply:
this is like keras thing: basically auto tracking relies on `__setattr__` being triggered. if `__setattr__` is not triggered then the variables / layers are actually not being tracked in auto tracker. When `__setattr__` is not properly triggered, all layers can still properly run but anything relies on autotracking being properly worked will break.

this is a slightly bigger thing in torch since torch relies on `torch.ModuleList` instead of plain dictionary for any collection of layers. with the keras3 logic (where a hook is added in torch layer when `__setattr__` is triggered to register modules) it is very important to ensure `__setattr__` is properly triggered.

Reply:
It is still unclear to me why this PR breaks `keras/src/models/sequential_test.py::SequentialTest::test_serialization` when using `self._layers.append(layer)`.

Shouldn't we have tracked these layers in `_maybe_rebuild`? It doesn't break in master branch.
I am worried that this discrepancy might break some models that use python containers (e.g. `list` and `dict`) for layer creation. Or is it already broken this way in the torch backend?

Reply:
> Or is it already broken this way in the torch backend?

yes - this is already broken on torch backend. you can try running below snippet in main branch v.s. the change

```
class StateAddingWrong(keras.layers.Layer):

    def build(self, _):
        self.layers = []
        self.layers.append(keras.layers.Dense(4, name="0"))
        self.layers.append(keras.layers.Dense(5, name="1"))
        self.layers.append(keras.layers.Dense(6, name="2"))

    def call(self, x):
        y = x
        for l in self.layers:
            y = l(y)
        return y


class StateAddingCorrect(keras.layers.Layer):

    def build(self, _):
        layers = []
        layers.append(keras.layers.Dense(4, name="0"))
        layers.append(keras.layers.Dense(5, name="1"))
        layers.append(keras.layers.Dense(6, name="2"))
        self.layers = layers

    def call(self, x):
        y = x
        for l in self.layers:
            y = l(y)
        return y


layer = StateAddingWrong(name='wrong')
layer(np.ones((2, 3)))
print(f"wrong tracked layers {len(layer._tracker.stored_ids['layers'])}")
print(f"wrong ", list(layer.named_modules()))

layer = StateAddingCorrect(name="correct")
layer(np.ones((2, 3)))
print(f"correct tracked layers {len(layer._tracker.stored_ids['layers'])}")
print(f"correct ", list(layer.named_modules()))
```

basically the `named_modules()` won't return layers appends to layers. my change fixes the case by using the pattern of not appending to layers. the most correct way to tackle this might be handling callbacks here: https://github.com/keras-team/keras/blob/master/keras/src/utils/tracking.py#L144-L147 @fchollet let me know if you have better ideas that could also sync the appending cases for torch backend.


> It is still unclear to me why this PR breaks keras/src/models/sequential_test.py::SequentialTest::test_serialization when using self._layers.append(layer)

so a little bit deeper rca:

the main error is on `_layers_{0, 1, 2}` attributes missing in deserialization test and i have found that the main issue is due to build & deserialization order.

here is the minimum reproducible test:

```
model = keras.Sequential(name="seq")
model.add(keras.layers.Dense(4, name="0"))
model.add(keras.layers.Dense(5, name="1"))
model.add(keras.layers.Dense(6, name="2"))
model.build((2, 3))

for m in model.named_modules():
    print(f"module {m}")
print("_layers_0" in dir(model))

config = model.get_config()
config_json = json.dumps(config, sort_keys=True, indent=4)
revived_model = keras.Sequential.from_config(config)

for m in model.named_modules():
    print(f"module {m}")
print("_layers_1" in dir(revived_model))
```

during build phase, `self._layers` is not set till `model.build` is called since https://github.com/keras-team/keras/blob/master/keras/src/models/sequential.py#L137-L145 conditions are not met. Thus, the layers will actually being called with all layers in the model with `model.build(shape)` call here: https://github.com/keras-team/keras/blob/master/keras/src/models/sequential.py#L179-L182. during deserialization phase, `InputLayer` is actually available so every `add` is rebuilding the model without resetting layers. the fix i have essentially forced resetting every time.

functionality wise all things are good since `Functional` in `Sequential` is actually the model being invoked and it has layers reset all the time here: https://github.com/keras-team/keras/blob/master/keras/src/models/functional.py#L140

Reply:
Thank you for the detailed explanation!

If this is the case, would it be more reasonable to put the logic in `_maybe_build`, since we might want to track these layers in both `add` and `pop`?

Actually, I think your solution of adding the logic in `track.py` is the best, but it might add some backend-specific code to it. This should wait for @fchollet 's call.

Reply:
> If this is the case, would it be more reasonable to put the logic in _maybe_build, since we might want to track these layers in both add and pop?

Agreed.

Reply:
@fchollet what's your thought on adding it to `tracking.py`? putting logics in `_maybe_build` can fix sequential specific issue but `append/pop` still won't be tracked by torch in current integration unless user triggers a `__setattr__`.

Options are:
1. fixing this by adding some backend specific callbacks in `tracking.py`
2. add a validation step in torch layer and warn user there are layers that not properly registered in torch.

Reply:
^ @fchollet let me know if you have any thoughts on above